From 60455e5ffe5129a592f356726565ab4c2ad1b87f Mon Sep 17 00:00:00 2001
From: Dong Ma <dong.ma@vexxhost.com>
Date: Mon, 15 Sep 2025 09:02:03 +0000
Subject: [PATCH] feat(dashboard): add microversion support for Nova live
 migration

- Include "live_migrate" in MICROVERSION_FEATURES with minimum version 2.30.
- Update `server_live_migrate` to use the appropriate microversion when calling
  Nova's live_migrate API.

Change-Id: Ief547b55f90ff2bfd9f97ea859b4511c631676ef
Closes-Bug: 1788367
Signed-off-by: Dong Ma <dong.ma@vexxhost.com>
---
 openstack_dashboard/api/microversions.py                       | 1 +
 openstack_dashboard/api/nova.py                                | 8 +++++---
 .../notes/nova-live-migrate-microversion-c3ccf731126b6e28.yaml | 6 ++++++
 3 files changed, 12 insertions(+), 3 deletions(-)
 create mode 100644 releasenotes/notes/nova-live-migrate-microversion-c3ccf731126b6e28.yaml

diff --git a/openstack_dashboard/api/microversions.py b/openstack_dashboard/api/microversions.py
index 556b85d..9322e33 100644
--- a/openstack_dashboard/api/microversions.py
+++ b/openstack_dashboard/api/microversions.py
@@ -39,6 +39,7 @@
         "key_type_list": ["2.9"],
         "rescue_instance_volume_based": ["2.87", "2.93"],
         "bdm2_volume_type": ["2.67", "2.96"],
+        "live_migrate": ["2.30"],
     },
     "cinder": {
         "groups": ["3.27", "3.43", "3.48", "3.58"],
diff --git a/openstack_dashboard/api/nova.py b/openstack_dashboard/api/nova.py
index 532446b..26e11b4 100644
--- a/openstack_dashboard/api/nova.py
+++ b/openstack_dashboard/api/nova.py
@@ -609,9 +609,11 @@
 @profiler.trace
 def server_live_migrate(request, instance_id, host, block_migration=False,
                         disk_over_commit=False):
-    _nova.novaclient(request).servers.live_migrate(instance_id, host,
-                                                   block_migration,
-                                                   disk_over_commit)
+    microversion = get_microversion(request, "live_migrate")
+    _nova.novaclient(request, version=microversion).servers.live_migrate(
+        instance_id, host,
+        block_migration,
+        disk_over_commit)


 @profiler.trace
diff --git a/releasenotes/notes/nova-live-migrate-microversion-c3ccf731126b6e28.yaml b/releasenotes/notes/nova-live-migrate-microversion-c3ccf731126b6e28.yaml
new file mode 100644
index 0000000..f5c369e
--- /dev/null
+++ b/releasenotes/notes/nova-live-migrate-microversion-c3ccf731126b6e28.yaml
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Horizon now supports Nova live migration with microversion 2.30.
+    The ``server_live_migrate`` API call will request the appropriate
+    microversion when available.
--
2.25.1
